File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.java
import java.util.*;
class Solution {
    public int solution(int[][] beginning, int[][] target) {
        int n = beginning.length;
        int m = beginning[0].length;
        
        int ans = Integer.MAX_VALUE;
        for(int row=0; row<(1<<n); row++) for(int col=0; col<(1<<m); col++){
            int[][] board = new int[n][m];
            for(int i=0;i<n;i++) for(int j=0;j<m;j++) board[i][j] = beginning[i][j];
            
            int cnt = 0;
            for(int i=0;i<n;i++){
                if((row&(1<<i))!=(1<<i)) continue;
                cnt++;
                for(int j=0;j<m;j++) board[i][j]^=1;
            }
            for(int i=0;i<m;i++){
                if((col&(1<<i))!=(1<<i)) continue;
                cnt++;
                for(int j=0;j<n;j++) board[j][i]^=1;
            }
            
            boolean flag = true;
            for(int i=0;i<n;i++) for(int j=0;j<m;j++){
                if(board[i][j]!=target[i][j]) flag = false;
            }
            
            if(flag){
                ans = Math.min(ans, cnt);
            }
        }

        
        return ans==Integer.MAX_VALUE ? -1: ans;
    }
}

File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
void colSwap(vector<vector<int>>& board, int c){
    for(int i=0;i<n;i++){
        board[i][c] ^= 1;
    }
}
void rowSwap(vector<vector<int>>& board, int r){
    for(int i=0;i<m;i++){
        board[r][i] ^= 1;
    }
}
bool chk(vector<vector<int>>& board, vector<vector<int>>& target){
    for(int i=0;i<n;i++) for(int j=0;j<m;j++){
        if(board[i][j]!=target[i][j]) return 0;
    }
    return 1;
}
int solution(vector<vector<int>> beginning, vector<vector<int>> target) {
    n = beginning.size();
    m = beginning[0].size();    
    
    int ans = 2e9;
    for(int brute = 0; brute < (1<<(n+m)); brute++){
        auto board = beginning;
        int cnt = 0;
        for(int i=0;i<n;i++){
            if((brute & (1<<i))==(1<<i)){
                cnt++;
                rowSwap(board,i);
            }
        }
        for(int i=n;i<n+m;i++){
            if((brute & (1<<i))==(1<<i)){
                cnt++;
                colSwap(board,i-n);
            }
        }
        if(chk(board, target)){
            ans = min(ans, cnt);
        }
        
    }
    
    return ans==2e9 ? -1 : ans;
}